src/devices/ec3k.c

```diff
@@ -0,0 +1,317 @@
+                }
+                // start and end of packet is marked by 6 ones surrounded by 0 (0x7e)
+                else if (onecount == 6) {
+                    packet    = !packet;
+                    packetpos = 0;
+                    recpos    = 0;
+                }
+                else {
+                    // reset state to re-sync
+                    packet    = 0;
+                    packetpos = 0;
+                    recpos    = 0;
+                }
+
+                onecount = 0;
+            }
+
+            if (packetpos >= DECODED_PAKET_LEN_BYTES) {
+                rc = ec3k_extract_fields(decoder, packetbuffer);
+
+                // reset state to re-sync
+                packet    = 0;
+                packetpos = 0;
+                recpos    = 0;
+                onecount  = 0;
+            }
+        }
+    }
+
+    return rc;
+}
+
+static int ec3k_extract_fields(r_device *const decoder, const uint8_t *packetbuffer)
+{
+    int rc = DECODE_FAIL_SANITY;
+
+    // decode received ec3k packet
+    uint16_t id             = unpack_nibbles(packetbuffer, 1, 4);
+    uint16_t time_total_low = unpack_nibbles(packetbuffer, 5, 4);
+    uint16_t pad_1          = unpack_nibbles(packetbuffer, 9, 4);
+    uint16_t time_on_low    = unpack_nibbles(packetbuffer, 13, 4);
+    uint32_t pad_2          = unpack_nibbles(packetbuffer, 17, 7);
+    uint32_t energy_low     = unpack_nibbles(packetbuffer, 24, 7);
+    double power_current    = unpack_nibbles(packetbuffer, 31, 4) / 10.0;
+    double power_max        = unpack_nibbles(packetbuffer, 35, 4) / 10.0;
+    // unknown? (seems to be used for internal calculations)
+    uint32_t energy2 = unpack_nibbles(packetbuffer, 39, 6);
+    // 						nibbles[45:59]
+    uint16_t time_total_high = unpack_nibbles(packetbuffer, 59, 3);
+    uint32_t pad_3           = unpack_nibbles(packetbuffer, 62, 5);
+    uint64_t energy_high     = (uint64_t)unpack_nibbles(packetbuffer, 67, 4) << 28;
+    uint16_t time_on_high    = unpack_nibbles(packetbuffer, 71, 3);
+    uint8_t reset_counter    = unpack_nibbles(packetbuffer, 74, 2);
+    uint8_t flags            = unpack_nibbles(packetbuffer, 76, 1);
+    uint8_t pad_4            = unpack_nibbles(packetbuffer, 77, 1);
+    uint16_t received_crc    = 0xffff ^ (unpack_nibbles(packetbuffer, 78, 2) | (unpack_nibbles(packetbuffer, 80, 2) << 8)); // little-endian
+    uint16_t calculated_crc  = calc_ec3k_crc(packetbuffer, DECODED_PAKET_LEN_BYTES - 2);
+
+    // convert to common units
+    uint64_t energy_Ws       = energy_high | energy_low;
+    const double energy_kWh  = energy_Ws / (1000.0 * 3600.0); // Ws to kWh
+    const double energy2_kWh = energy2 / (1000.0 * 3600.0);   // Ws to kWh
+    uint32_t time_total      = (uint32_t)time_total_low | ((uint32_t)time_total_high << 16);
+    uint32_t time_on         = (uint32_t)time_on_low | ((uint32_t)time_on_high << 16);
+
+    if (pad_1 == 0 && pad_2 == 0 && pad_3 == 0 && pad_4 == 0) {
```

Comment:
Can you flatten this / early return? Pull up the else, e.g. `if (pad != 0 || ...) { decoder_log(); return DECODE_FAIL_SANITY;}`